### ISSUE TITLE
$table_prefix is a wordpress global that is supposed to be overwritten.

### ISSUE THREAD

**Issue by LC43** [WordPress-Coding-Standards/WordPress-Coding-Standards#1639]
## Bug Description
$table_prefix is WordPress global that shouldn't be sniffed. It has to be instanciated in wp-config.php, when using a custom prefix.

## Minimal Code Snippet

```php
$table_prefix = 'wp_;
```
## Error Code

> Overriding WordPress globals is prohibited. Found assignment to $table_prefix 

Sniff: WordPress.WP.GlobalVariablesOverride.OverrideProhibited

## Environment

| Question               | Answer
| ------------------------| -------
| PHP version             | 7.3.0
| PHP_CodeSniffer version | 3.3.2
| WPCS version            | 2.0.0
| WPCS install type       | git clone

## Tested Against `develop` branch? 
yes:
- [x] I have verified the issue still exists in the `develop` branch of WPCS.

**Comment by jrfnl** [comment]
> $table_prefix is WordPress global that shouldn't be sniffed. It has to be instanciated in wp-config.php, when using a custom prefix.

Ok, so there are two issues in this statement:
1. I agree that it has to be instantiated in `wp-config.php` and that if it's found in that file, it shouldn't be flagged.
2. `$table_prefix` most definitely should be flagged if it's overwritten by a plugin or theme or other userland code as that would make your data unreachable.

The `WordPress.WP.GlobalVariablesOverride` sniff is contained within the `WordPress` and `WordPress-Extra` ruleset, however, so what is going on that you're scanning `wp-config.php` (a core file) with the `WordPress-Extra` ruleset in the first place ?

**Comment by LC43** [comment]
Hi @jrfnl 

1. great!

2. Fair point.

Our structure has wp in a custom dir `/wp/`, but wp-config.php sits in the root. 

We exclude `/wp/`, but wp-config.php isn't part of the core so it sits with the rest of the files in root dir.

I assume I should instead just prevent WPCS to sniff this line?

**Comment by jrfnl** [comment]
> I assume I should instead just prevent WPCS to sniff this line?

@LC43 Well, probably just for this sniff. Adding the below to your ruleset should do the trick.

```xml
<rule ref="WordPress.WP.GlobalVariablesOverride">
    <exclude-pattern>/wp-config\.php$</exclude-pattern>
</rule>
```

**Comment by LC43** [comment]
Thank you so much @jrfnl !

sorry for the noise, closing :)

**Comment by jrfnl** [comment]
No worries, I presume that means the snippet I gave you solved your problem ?

**Comment by LC43** [comment]
flawlessly!

**Comment by nbordeau** [comment]
I just added the following code to our main Wordpress `ruleset.xml`:

```
	<!--
	Added to suppress error in wp-config for overriding $table_prefix
	https://github.com/WordPress/WordPress-Coding-Standards/issues/1639
	-->
	<rule ref="WordPress.WP.GlobalVariablesOverride">
            <exclude-pattern>/wp-config\.php$</exclude-pattern>
	</rule>
```

Is this something that should be added by default to suppress that error?

**Comment by jrfnl** [comment]
@nbordeau Not really as `wp-config.php` is a WP Core file so would normally not be scanned by that sniff anyway.

Another thing is, that adding that snippet to WPCS would also suppress other notifications from that sniff for any file called `wp-config.php` and if a plugin/theme wanted to be malicious, it could of course try to abuse that.

So I'd rather - if people include `wp-config.php` in their scans in the first place - that they add this snippet themselves after examining the file to make sure it's the core file, than that WPCS would blanket ignore the file.

**Comment by nbordeau** [comment]
Great note @jrfnl. Is there a way that we can limit the sniffer to only exclude that for the main wp-config.php file in the root of our projects?

**Comment by jrfnl** [comment]
@nbordeau Yes, you can. You'd need to use something along the lines of `<exclude-pattern type="relative">^/wp-config\.php$</exclude-pattern>`.

You can find more information about that syntax in the PHPCS wiki: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-Ruleset
